README.md
# RaidCore


RaidCore/CHANGES.txt
tag 9f2a93eb0ec8fe636e8af474d6a5462cd3018b4f 2.0
Author:	NielsH <github@nuvini.com>
Date:	Mon May 11 21:16:39 2015 +0200

Release 2.0

commit ef19ef4ca1ba829e5c1473523f04a8fbe062d8dd
Author: NielsH <github@nuvini.com>
Date:   Mon May 11 21:11:27 2015 +0200

    Release 2.0
    
    Updated AddonVersion string to 15051101

commit ef19ef4ca1ba829e5c1473523f04a8fbe062d8dd
Author: NielsH <github@nuvini.com>
Date:   Mon May 11 21:11:27 2015 +0200

    Release 2.0
    
    Updated AddonVersion string to 15051101

commit 47855c9430ed902ce164570ae4fb7ad10afca06d
Author: NielsH <github@nuvini.com>
Date:   Mon May 11 21:03:18 2015 +0200

    Fix typo in ohmna, clean up some dead code, refs #68
    
    This commit fixes a typo for Ohmna. This might be the cause for the
    pillar timer not working, but still needs testing if it's fully fixed
    with this.
    Also removes call to StopScan() since we no longer use this.

commit 47855c9430ed902ce164570ae4fb7ad10afca06d
Author: NielsH <github@nuvini.com>
Date:   Mon May 11 21:03:18 2015 +0200

    Fix typo in ohmna, clean up some dead code, refs #68
    
    This commit fixes a typo for Ohmna. This might be the cause for the
    pillar timer not working, but still needs testing if it's fully fixed
    with this.
    Also removes call to StopScan() since we no longer use this.

commit 652a29f9ca0cf90f46fa0ff492440c71d6158b5f
Author: Imm├⌐ sama <immemor@users.noreply.github.com>
Date:   Mon May 11 19:26:51 2015 +0200

    Encounters/Y83: Add Prime Evolutionary Operant encounter, refs #91
    
    First feedback in easy mode.

commit 652a29f9ca0cf90f46fa0ff492440c71d6158b5f
Author: Imm├⌐ sama <immemor@users.noreply.github.com>
Date:   Mon May 11 19:26:51 2015 +0200

    Encounters/Y83: Add Prime Evolutionary Operant encounter, refs #91
    
    First feedback in easy mode.

commit 99c6ce3e923b60135969834f26e91a5e022ca0a1
Author: Imm├⌐ sama <immemor@users.noreply.github.com>
Date:   Mon May 11 02:38:15 2015 +0200

    Encounters/DS: Update LogicLife, refs #35
    
     - Replace buff identification string by spellId, refs #35
     - Update french language
     - delete useless function which is 'OnReset'.

commit 641dcddcbfc6ec027697df610af1a586279d84c0
Author: NielsH <github@nuvini.com>
Date:   Sun May 10 23:04:55 2015 +0200

    Fix call to StopBar() with incorrect key
    
    This commit fixes an issue with Kuralak where the StopBar() parameter
    was the translation string, and not the actual key we used earlier when
    adding the bar.

commit 641dcddcbfc6ec027697df610af1a586279d84c0
Author: NielsH <github@nuvini.com>
Date:   Sun May 10 23:04:55 2015 +0200

    Fix call to StopBar() with incorrect key
    
    This commit fixes an issue with Kuralak where the StopBar() parameter
    was the translation string, and not the actual key we used earlier when
    adding the bar.

commit 99c6ce3e923b60135969834f26e91a5e022ca0a1
Author: Imm├⌐ sama <immemor@users.noreply.github.com>
Date:   Mon May 11 02:38:15 2015 +0200

    Encounters/DS: Update LogicLife, refs #35
    
     - Replace buff identification string by spellId, refs #35
     - Update french language
     - delete useless function which is 'OnReset'.

commit 641dcddcbfc6ec027697df610af1a586279d84c0
Author: NielsH <github@nuvini.com>
Date:   Sun May 10 23:04:55 2015 +0200

    Fix call to StopBar() with incorrect key
    
    This commit fixes an issue with Kuralak where the StopBar() parameter
    was the translation string, and not the actual key we used earlier when
    adding the bar.

commit 183d227d152e27861e0aa3052a1001cb92a0c60c
Author: NielsH <github@nuvini.com>
Date:   Sun May 10 21:25:19 2015 +0200

    Fix world marker displayed on units not on screen, closes #90
    
    Previously we had already fixed the issue where world markers were
    displayed on the wrong location on the screen when the position
    coordinates were not on our screen. This commit extends that
    functionality to marking of units as well.

commit ddd22500d4a78092e0cce2b5ab5ae385fceda20c
Author: NielsH <github@nuvini.com>
Date:   Sat May 9 17:01:14 2015 +0200

    Add option to decline summons, refs #100
    
    Raid leaders and assistants can force people to accept pending summons
    via /raidc summon.
    This will allow people to override that option in the settings, so
    summons are no longer automatically accepted.

commit ddd22500d4a78092e0cce2b5ab5ae385fceda20c
Author: NielsH <github@nuvini.com>
Date:   Sat May 9 17:01:14 2015 +0200

    Add option to decline summons, refs #100
    
    Raid leaders and assistants can force people to accept pending summons
    via /raidc summon.
    This will allow people to override that option in the settings, so
    summons are no longer automatically accepted.

commit 79f1bf79ace7cb7eb0a23fd0697bd5851c403f47
Author: NielsH <github@nuvini.com>
Date:   Sat May 9 16:28:58 2015 +0200

    Fix / Improve System Daemons World Markers
    
    System Daemons still had calls to the old method of World Markers so
    they didn't work at all. This commit fixes that issue. It also
    improves/reduces the code a bit by placing the markers via a loop
    instead of separate calls.

commit f67ed90d67484014e6c2d19703af70defed46fa3
Author: NielsH <github@nuvini.com>
Date:   Sat May 9 14:44:31 2015 +0200

    Fix an issue that prevented OnReset() from working
    
    The RAID_WIPE event was fired after the encounter was disabled, at that
    point it doesn't work so OnReset() was never called. Since this is
    required for some fights, I moved the event handler so it fires before
    disabling the module.
    
    This was an issue specifically on Avatus, where it only worked the first
    pull, until another reloadui happened.

commit f67ed90d67484014e6c2d19703af70defed46fa3
Author: NielsH <github@nuvini.com>
Date:   Sat May 9 14:44:31 2015 +0200

    Fix an issue that prevented OnReset() from working
    

RaidCore/Encounters/DS/EpAirLife.lua
--------------------------------------------------------------------------------
-- Module Declaration
--

local core = Apollo.GetPackage("Gemini:Addon-1.1").tPackage:GetAddon("RaidCore")

local mod = core:NewEncounter("EpAirLife", 52, 98, 119)
if not mod then return end

mod:RegisterTrigMob("ALL", { "Aileron", "Visceralus" })
mod:RegisterEnglishLocale({
    -- Unit names.
    ["Visceralus"] = "Visceralus",
    ["Aileron"] = "Aileron",
    ["Wild Brambles"] = "Wild Brambles",
    ["[DS] e395 - Air - Tornado"] = "[DS] e395 - Air - Tornado",
    ["Life Force"] = "Life Force",
    ["Lifekeeper"] = "Lifekeeper",
    -- Datachron messages.
    -- Cast.
    ["Blinding Light"] = "Blinding Light",
    -- Bar and messages.
    ["TWIRL ON YOU!"] = "TWIRL ON YOU!",
    ["Thorns"] = "Thorns",
    ["Twirl"] = "Twirl",
    ["Midphase ending"] = "Midphase ending",
    ["Middle Phase"] = "Middle Phase",
    ["Next Healing Tree"] = "Next Healing Tree",
    ["No-Healing Debuff!"] = "No-Healing Debuff!",
    ["NO HEAL DEBUFF"] = "NO HEAL\nDEBUFF",
    ["Lightning"] = "Lightning",
    ["Lightning on YOU"] = "Lightning on YOU",
    ["Recently Saved!"] = "Recently Saved!",
})
mod:RegisterFrenchLocale({
    -- Unit names.
    ["Visceralus"] = "Visceralus",
    ["Aileron"] = "Ventemort",
    ["Wild Brambles"] = "Ronces sauvages",
    --["[DS] e395 - Air - Tornado"] = "[DS] e395 - Air - Tornado", -- TODO: French translation missing !!!!
    ["Life Force"] = "Force vitale",
    ["Lifekeeper"] = "Garde-vie",
    -- Datachron messages.
    -- Cast.
    ["Blinding Light"] = "Lumière aveuglante",
    -- Bar and messages.
    --["TWIRL ON YOU!"] = "TWIRL ON YOU!", -- TODO: French translation missing !!!!
    ["Thorns"] = "Épines",
    ["Twirl"] = "Tournoiement",
    --["Midphase ending"] = "Midphase ending", -- TODO: French translation missing !!!!
    --["Middle Phase"] = "Middle Phase", -- TODO: French translation missing !!!!
    --["Next Healing Tree"] = "Next Healing Tree", -- TODO: French translation missing !!!!
    --["No-Healing Debuff!"] = "No-Healing Debuff!", -- TODO: French translation missing !!!!
    --["NO HEAL DEBUFF"] = "NO HEAL\nDEBUFF", -- TODO: French translation missing !!!!
    ["Lightning"] = "Foudre",
    --["Lightning on YOU"] = "Lightning on YOU", -- TODO: French translation missing !!!!
    --["Recently Saved!"] = "Recently Saved!", -- TODO: French translation missing !!!!
})
mod:RegisterGermanLocale({
    -- Unit names.
    ["Visceralus"] = "Viszeralus",
    ["Aileron"] = "Aileron",
    ["Wild Brambles"] = "Wilde Brombeeren",
    --["[DS] e395 - Air - Tornado"] = "[DS] e395 - Air - Tornado", -- TODO: German translation missing !!!!
    ["Life Force"] = "Lebenskraft",
    ["Lifekeeper"] = "Lebensbewahrer",
    -- Datachron messages.
    -- Cast.
    ["Blinding Light"] = "Blendendes Licht",
    -- Bar and messages.
    --["TWIRL ON YOU!"] = "TWIRL ON YOU!", -- TODO: German translation missing !!!!
    ["Thorns"] = "Dornen",
    ["Twirl"] = "Wirbel",
    --["Midphase ending"] = "Midphase ending", -- TODO: German translation missing !!!!
    --["Middle Phase"] = "Middle Phase", -- TODO: German translation missing !!!!
    --["Next Healing Tree"] = "Next Healing Tree", -- TODO: German translation missing !!!!
    --["No-Healing Debuff!"] = "No-Healing Debuff!", -- TODO: German translation missing !!!!
    --["NO HEAL DEBUFF"] = "NO HEAL\nDEBUFF", -- TODO: German translation missing !!!!
    ["Lightning"] = "Blitz",
    --["Lightning on YOU"] = "Lightning on YOU", -- TODO: German translation missing !!!!
    --["Recently Saved!"] = "Recently Saved!", -- TODO: German translation missing !!!!
})

-- Tracking Blinding Light and Aileron knockback seems too random to display on timers.

--------------------------------------------------------------------------------
-- Locals
--

local last_thorns = 0
local last_twirl = 0
local midphase = false
local myName
local CheckTwirlTimer = nil
local twirl_units = {}
local twirlCount = 0
local lightningCount = 0

--------------------------------------------------------------------------------
-- Initialization
--
function mod:OnBossEnable()
    Print(("Module %s loaded"):format(mod.ModuleName))
    Apollo.RegisterEventHandler("RC_UnitCreated", "OnUnitCreated", self)
    Apollo.RegisterEventHandler("RC_UnitDestroyed", "OnUnitDestroyed", self)
    Apollo.RegisterEventHandler("RC_UnitStateChanged", "OnUnitStateChanged", self)
    Apollo.RegisterEventHandler("SPELL_CAST_START", "OnSpellCastStart", self)
    Apollo.RegisterEventHandler("DEBUFF_APPLIED", "OnDebuffApplied", self)
    Apollo.RegisterEventHandler("DEBUFF_REMOVED", "OnDebuffRemoved", self)
    Apollo.RegisterEventHandler("RAID_WIPE", "OnReset", self)
end

--------------------------------------------------------------------------------
-- Event Handlers
--

function mod:OnReset()
    last_thorns = 0
    last_twirl = 0
    midphase = false
    if CheckTwirlTimer then
        self:CancelTimer(CheckTwirlTimer)
    end
    twirl_units = {}
    twirlCount = 0
	lightningCount = 0

    core:StopBar("THORN")
    core:StopBar("MIDEND")
    core:StopBar("MIDPHASE")
    core:StopBar("TWIRL")
end

function mod:OnUnitCreated(unit, sName)
    local eventTime = GameLib.GetGameTime()
    if sName == self.L["Wild Brambles"] and eventTime > last_thorns + 1 and eventTime + 16 < midphase_start then
        last_thorns = eventTime
        twirlCount = twirlCount + 1
        core:AddBar("THORN", self.L["Thorns"], 15)
        if twirlCount == 1 then
            core:AddBar("TWIRL", self.L["Twirl"], 15)
        elseif twirlCount % 2 == 1 then
            core:AddBar("TWIRL", self.L["Twirl"], 15)
        end
    elseif not midphase and sName == self.L["[DS] e395 - Air - Tornado"] then
        midphase = true
        twirlCount = 0
        midphase_start = eventTime + 115
        core:AddBar("MIDEND", self.L["Midphase ending"], 35)
        core:AddBar("THORN", self.L["Thorns"], 35)
        core:AddBar("Lifekeep", self.L["Next Healing Tree"], 35)
    elseif sName == self.L["Life Force"] and mod:GetSetting("LineLifeOrbs") then
        core:AddPixie(unit:GetId(), 2, unit, nil, "Blue", 10, 40, 0)
    elseif sName == self.L["Lifekeeper"] then
        if mod:GetSetting("LineHealingTrees") then
            core:AddPixie(unit:GetId(), 1, GameLib.GetPlayerUnit(), unit, "Yellow", 5, 10, 10)
        end
        core:AddUnit(unit)
        core:AddBar("Lifekeep", self.L["Next Healing Tree"], 30, mod:GetSetting("SoundHealingTree"))
    end
end

function mod:OnUnitDestroyed(unit, sName)
    local eventTime = GameLib.GetGameTime()
    if midphase and sName == self.L["[DS] e395 - Air - Tornado"] then
        midphase = false
        core:AddBar("MIDPHASE", self.L["Middle Phase"], 90, mod:GetSetting("SoundMidphase"))
    elseif sName == self.L["Life Force"] then
        core:DropPixie(unit:GetId())
    elseif sName == self.L["Lifekeeper"] then
        core:DropPixie(unit:GetId())
    end
end

function mod:OnDebuffApplied(unitName, splId, unit)
    local eventTime = GameLib.GetGameTime()
    local splName = GameLib.GetSpell(splId):GetName()
    if splId == 70440 then -- Twirl
        if unitName == myName and mod:GetSetting("OtherTwirlWarning") then
            core:AddMsg("TWIRL", self.L["TWIRL ON YOU!"], 5, mod:GetSetting("SoundTwirl", "Inferno"))
        end

        if mod:GetSetting("OtherTwirlPlayerMarkers") then
            core:MarkUnit(unit, nil, self.L["Twirl"]:upper())
        end
        core:AddUnit(unit)
        twirl_units[unitName] = unit
        if not CheckTwirlTimer then
            CheckTwirlTimer = self:ScheduleRepeatingTimer("CheckTwirlTimer", 1)
        end
    elseif splName == "Life Force Shackle" then
        if mod:GetSetting("OtherNoHealDebuffPlayerMarkers") then
            core:MarkUnit(unit, nil, self.L["NO HEAL DEBUFF"])
        end
        if unitName == strMyName and mod:GetSetting("OtherNoHealDebuff") then
            core:AddMsg("NOHEAL", self.L["No-Healing Debuff!"], 5, mod:GetSetting("SoundNoHealDebuff", "Alarm"))
        end
    elseif splName == "Lightning Strike" then
        if unitName == strMyName then
            core:AddMsg("LIGHTNING", self.L["Lightning on YOU"], 5, mod:GetSetting("SoundLightning", "RunAway"))
        end

		--modifications:

		lightningCount = lightningCount + 1

		if mod:GetSetting("OtherLightningMarkers") then
			if lightningCount <= 2 then
            	core:MarkUnit(unit, nil, "Lightning\n1")
			else
				core:MarkUnit(unit, nil, "Lightning\n2")
			end
        end
		
		if lightningCount >= 4 then
			lightningSet = lightningSet + 1
			lightningCount = 0
		end
    end
end

function mod:OnDebuffRemoved(unitName, splId, unit)
    local splName = GameLib.GetSpell(splId):GetName()
    if splId == 70440 then
        core:RemoveUnit(unit:GetId())
    elseif splName == "Life Force Shackle" then
        core:DropMark(unit:GetId())
    elseif splName == "Lightning Strike" then
        core:DropMark(unit:GetId())
    end
end

function mod:OnSpellCastStart(unitName, castName, unit)
    local eventTime = GameLib.GetGameTime()
    if unitName == self.L["Visceralus"] and castName == self.L["Blinding Light"] and mod:GetSetting("OtherBlindingLight") then
        local playerUnit = GameLib.GetPlayerUnit()
        if self:GetDistanceBetweenUnits(unit, playerUnit) < 33 then
            core:AddMsg("BLIND", self.L["Blinding Light"], 5, mod:GetSetting("SoundBlindingLight", "Beware"))
        end
    end
end

function mod:CheckTwirlTimer()
    for unitName, unit in pairs(twirl_units) do
        if unit and unit:GetBuffs() then
            local bUnitHasTwirl = false
            local debuffs = unit:GetBuffs().arHarmful
            for _, debuff in pairs(debuffs) do
                if debuff.splEffect:GetId() == 70440 then -- the Twirl ability
                    bUnitHasTwirl = true
                end
            end
            if not bUnitHasTwirl then
                -- else, if the debuff is no longer present, no need to track anymore.
                core:DropMark(unit:GetId())
                core:RemoveUnit(unit:GetId())
                twirl_units[unitName] = nil
            end
        end
    end
end

function mod:OnUnitStateChanged(unit, bInCombat, sName)
    if unit:GetType() == "NonPlayer" and bInCombat then
        local eventTime = GameLib.GetGameTime()
        local playerUnit = GameLib.GetPlayerUnit()
        myName = playerUnit:GetName()

        if sName == self.L["Aileron"] then
            core:AddUnit(unit)
            if mod:GetSetting("LineCleaveAileron") then
                core:AddPixie(unit:GetId(), 2, unit, nil, "Red", 10, 30, 0)
            end
        elseif sName == self.L["Visceralus"] then
            core:AddUnit(unit)
            core:WatchUnit(unit)
            core:RaidDebuff()

            last_thorns = 0
            last_twirl = 0
            twirl_units = {}
            CheckTwirlTimer = nil
            midphase = false
            midphase_start = eventTime + 90
            twirlCount = 0

            core:AddBar("MIDPHASE", self.L["Middle Phase"], 90, mod:GetSetting("SoundMidphase"))
            core:AddBar("THORN", self.L["Thorns"], 20)
        end
    end
end
